### PR TITLE
Refactored URL to complete-validate-run pattern

### LINKED ISSUE #1193: Refactor url command to follow complete-validate-run pattern
more info in #1142

### PR DESCRIPTION
**Pull request by surajnarwade**
## What is the purpose of this change? What does it change?

Refactoring url to `complete-validate-run` pattern

## Was the change discussed in an issue?
fixes #1193 

## How to test changes?

* Check CI status to :green_heart:  and run all url related commands :smile:

### CODE REVIEW

**Comment by geoand** [comment]
@surajnarwade Could you fix the names in order to keep CodeClimate happy?

Thanks

File: pkg/odo/cli/url/url.go [hunk 1]
```diff
@@ -2,200 +2,45 @@ package url
...
-	urlCreateCmd.Flags().IntVarP(&urlPort, "port", "", -1, "port number for the url of the component, required in case of components which expose more than one service port")
-	urlCreateCmd.Flags().BoolVar(&urlOpenFlag, "open", false, "open the created link with your default browser")
-
-	urlDeleteCmd.Flags().BoolVarP(&urlForceDeleteFlag, "force", "f", false, "Delete url without prompting")
-
-	urlCmd.AddCommand(urlListCmd)
-	urlCmd.AddCommand(urlDeleteCmd)
-	urlCmd.AddCommand(urlCreateCmd)
+func NewCmdURL(name, fullName string) *cobra.Command {
+	urlCreateCmd := NewCmdUrlCreate(createRecommendedCommandName, fullName+" "+createRecommendedCommandName)
```
Comment (metacosm) [inline pkg/odo/cli/url/url.go]: Use `util.GetFullName` here and for `urlDeleteCmd` and `urlListCmd`.

File: pkg/odo/cli/url/delete.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+		Run: func(cmd *cobra.Command, args []string) {
+			odoutil.LogErrorAndExit(o.Complete(name, cmd, args), "")
+			odoutil.LogErrorAndExit(o.Validate(), "")
+			odoutil.LogErrorAndExit(o.Run(), "")
+		},
+		Example: fmt.Sprintf(urlDeleteExample, fullName),
+	}
+	urlDeleteCmd.Flags().BoolVarP(&o.urlForceDeleteFlag, "force", "f", false, "Delete url without prompting")
+
+	completion.RegisterCommandHandler(urlDeleteCmd, completion.URLCompletionHandler)
```
Comment (metacosm) [inline pkg/odo/cli/url/delete.go]: 👍

File: pkg/odo/cli/url/create.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+// NewUrlCreateOptions creates a new UrlCreateOptions instance
+func NewUrlCreateOptions() *UrlCreateOptions {
+	return &UrlCreateOptions{}
+}
+
+// Complete completes UrlCreateOptions after they've been Created
+func (o *UrlCreateOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	o.componentPort, err = url.GetValidPortNumber(o.Client, o.urlPort, o.Component(), o.Application)
```
Comment (metacosm) [inline pkg/odo/cli/url/create.go]: You need to initialize `o.Context` somehow before being able to use its methods, most likely using `o.Context = genericclioptions.NewContext(cmd)` before this line.

File: pkg/odo/cli/url/delete.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+// NewUrlDeleteOptions creates a new UrlDeleteOptions instance
+func NewUrlDeleteOptions() *UrlDeleteOptions {
+	return &UrlDeleteOptions{}
+}
+
+// Complete completes UrlDeleteOptions after they've been Deleted
+func (o *UrlDeleteOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	o.urlName = args[0]
```
Comment (metacosm) [inline pkg/odo/cli/url/delete.go]: Same as for `create`: you need to initialize the context here.

File: pkg/odo/cli/url/list.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+// NewUrlCreateOptions creates a new UrlCreateOptions instance
+func NewUrlListOptions() *UrlListOptions {
+	return &UrlListOptions{}
+}
+
+// Complete completes UrlListOptions after they've been Listed
+func (o *UrlListOptions) Complete(name string, cmd *cobra.Command, args []string) (err error) {
+	o.Context = genericclioptions.NewContext(cmd)
```
Comment (metacosm) [inline pkg/odo/cli/url/list.go]: 👍

**Comment by metacosm** [comment]
Also, please link the associated issue in the PR description. Thanks!

File: pkg/odo/cli/url/url.go [hunk 6]
```diff
@@ -2,200 +2,45 @@ package url
...
-	urlCreateCmd.Flags().IntVarP(&urlPort, "port", "", -1, "port number for the url of the component, required in case of components which expose more than one service port")
-	urlCreateCmd.Flags().BoolVar(&urlOpenFlag, "open", false, "open the created link with your default browser")
-
-	urlDeleteCmd.Flags().BoolVarP(&urlForceDeleteFlag, "force", "f", false, "Delete url without prompting")
-
-	urlCmd.AddCommand(urlListCmd)
-	urlCmd.AddCommand(urlDeleteCmd)
-	urlCmd.AddCommand(urlCreateCmd)
+func NewCmdURL(name, fullName string) *cobra.Command {
+	urlCreateCmd := NewCmdURLCreate(createRecommendedCommandName, odoutil.GetFullName(fullName, createRecommendedCommandName))
```
Comment (metacosm) [inline pkg/odo/cli/url/url.go]: 👍

File: pkg/odo/cli/url/create.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		o.urlName = args[0]
+	default:
+		return fmt.Errorf("unable to get component")
+	}
+	return
+}
+
+// Validate validates the UrlCreateOptions based on completed values
+func (o *URLCreateOptions) Validate() (err error) {
+	exists, err := url.Exists(o.Client, o.urlName, "", o.Application)
```
Comment (metacosm) [inline pkg/odo/cli/url/create.go]: I understand that this was like this before this refactoring but why aren't we passing the component name here?

File: pkg/odo/cli/url/delete.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+		fmt.Scanln(&confirmDeletion)
+	}
+
+	if strings.ToLower(confirmDeletion) == "y" {
+
+		err = url.Delete(o.Client, o.urlName, o.Application)
+		odoutil.LogErrorAndExit(err, "")
+		log.Infof("Deleted URL: %v", o.urlName)
+	} else {
+		log.Errorf("Aborting deletion of url: %v", o.urlName)
```
Comment (metacosm) [inline pkg/odo/cli/url/delete.go]: return the error instead of exiting here

File: pkg/odo/cli/url/delete.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+		confirmDeletion = "y"
+	} else {
+		log.Askf("Are you sure you want to delete the url %v? [y/N]: ", o.urlName)
+		fmt.Scanln(&confirmDeletion)
+	}
+
+	if strings.ToLower(confirmDeletion) == "y" {
+
+		err = url.Delete(o.Client, o.urlName, o.Application)
+		odoutil.LogErrorAndExit(err, "")
```
Comment (metacosm) [inline pkg/odo/cli/url/delete.go]: return the error instead of exiting here

**Comment by metacosm** [comment]
Please, do not squash your commits and/or rebase your code until all the comments have been addressed. This makes things difficult to review because we cannot track down whether comments have been addressed easily.

**Pull request merged**